durabletask/internal/grpc_interceptor.py

```diff
@@ -0,0 +1,63 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+from collections import namedtuple
+from typing import List, Tuple
+
+import grpc
+
+
+class _ClientCallDetails(
+        namedtuple(
+            '_ClientCallDetails',
+            ['method', 'timeout', 'metadata', 'credentials', 'wait_for_ready', 'compression']),
+        grpc.ClientCallDetails):
+    """This is an implementation of the ClientCallDetails interface needed for interceptors.
+    This class takes six named values and inherits the ClientCallDetails from grpc package.
+    This class encloses the values that describe a RPC to be invoked.
+    """
+    pass
+
+
+class DefaultClientInterceptorImpl (
+    grpc.UnaryUnaryClientInterceptor, grpc.UnaryStreamClientInterceptor,
+    grpc.StreamUnaryClientInterceptor, grpc.StreamStreamClientInterceptor):
+    """The class implements a UnaryUnaryClientInterceptor, UnaryStreamClientInterceptor,
+    StreamUnaryClientInterceptor and StreamStreamClientInterceptor from grpc to add an 
+    interceptor to add additional headers to all calls as needed."""
+
+    def __init__(self, metadata: List[Tuple[str, str]]):
+        super().__init__()
+        self._metadata = metadata
+
+    def _intercept_call(
+                self, client_call_details: _ClientCallDetails) -> grpc.ClientCallDetails:
+        """Internal intercept_call implementation which adds metadata to grpc metadata in the RPC
+            call details."""
+        if client_call_details.metadata is not None:
+            metadata = list(client_call_details.metadata)
+        else:
+            metadata = []
+        if self._metadata is not None:
```

Comment:
This check should be moved to the very beginning of the function. We don't want to allocate anything if `self._metadata` is not initialized.

```python
def _intercept_call(...):
    if self._metadata is None:
        return client_call_details

    if client_call_details.metadata is not None:
        ...
```

Reply:
incorporated